src/Disks/DiskCache.h

```diff
@@ -0,0 +1,167 @@
+#pragma once
+
+#include "DiskLocal.h"
+#include "DiskCacheMetadata.h"
+#include <unordered_map>
+#include <list>
+#include <functional>
+
+namespace DB
+{
+
+class DiskCachePolicy
+{
+public:
+    virtual ~DiskCachePolicy() {}
+
+    struct CachePart
+    {
+        enum class CachePartType
+        {
+            CACHED,
+            ABSENT,
+            ABSENT_NO_CACHE,
+            EMPTY,
+        };
+        CachePart(size_t offset_, size_t size_) : offset(offset_), size(size_) {}
+        size_t offset;
+        size_t size;
+        CachePartType type = CachePartType::CACHED;
+    };
+
+    typedef std::list<CachePart> CachePartList;
+
+    /// fing object parts in cache
+    /// wait if some part is in progress
+    /// return next part to read or download
+    /// returned part marked as in progress
+    virtual CachePart find(const String & key, size_t offset, size_t size) = 0;
+
+    /// reserve some size in cache, with eviction old records if needs
+    /// returns list of keys to remove from cache
+    virtual std::list<std::pair<String, size_t>> reserve(size_t size) = 0;
+
+    /// insert file in cache index
+    /// returns list of keys to remove from cache (as reserve)
+    virtual std::list<std::pair<String, size_t>> add(const String & key, size_t file_size,
+        size_t offset, size_t size, bool restore) = 0;
+
+    /// call on download error
+    virtual void error(const String & key, size_t offset) = 0;
+
+    /// remove file from cache index
+    virtual CachePartList remove(const String & key) = 0;
+
+    /// remove one record from cache index
+    virtual void remove(const String & key, size_t offset, size_t size) = 0;
+
+    /// mark record as used (for LRU, etc.)
+    virtual void read(const String & key, size_t offset, size_t size) = 0;
+};
+
+class DiskCacheLRUPolicy : public DiskCachePolicy
```

Comment:
Can we add some more policies like ARC or 2Q? If no, then this interface is not so good.

Reply:
BTW, we already have the best performance LRU cache in our codebase. Try to adapt it here.

Reply:
Almost: https://github.com/ClickHouse/ClickHouse/pull/22327

Reply:
1. Yes, it's possible to use other policies just by make other implemetation.
2. Main differences from "classic" cache
  i) Cache limit not only by cached elements but by size of elements (it's possible to drop several elements from cache on insert one large)
  ii) Find not by exact match key but by range.
  iii) Wait on lock during find to avoid parallel downloading the same parts of remote object.
So this cache is more complificated and I think thad hard to reuse other existing implementation.